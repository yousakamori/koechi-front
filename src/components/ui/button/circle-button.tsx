import clsx from 'clsx';
import React from 'react';
// ___________________________________________________________________________
//
const mapSizeClasses = {
  sm: ['p-1', 'w-7', 'h-7'],
  md: ['p-1', 'w-9', 'h-9'],
  lg: ['p-1', 'w-12', 'h-12'],
};

const mapIconSize = {
  sm: ['text-base'],
  md: ['text-xl'],
  lg: ['text-2xl'],
};

const mapColorClasses = {
  primary: ['text-primary-400', 'hover:text-primary-500'],
  secondary: ['text-secondary-500', 'hover:text-secondary-800'],
  error: ['text-error-400', 'hover:text-error-500'],
};
// ___________________________________________________________________________
//
export type BaseCircleButtonProps = {
  variant?: 'ghost' | 'none';
  size?: keyof typeof mapSizeClasses;
  color?: keyof typeof mapColorClasses;
  disabled?: boolean;
  transition?: boolean;
  tooltip?: boolean;
};

export type AnchorCircleButtonProps = BaseCircleButtonProps &
  Omit<React.AnchorHTMLAttributes<HTMLAnchorElement>, keyof BaseCircleButtonProps>;

export type NativeCircleButtonProps = BaseCircleButtonProps &
  Omit<React.ButtonHTMLAttributes<HTMLButtonElement>, keyof BaseCircleButtonProps>;

export type CircleButtonProps = AnchorCircleButtonProps & NativeCircleButtonProps;
// ___________________________________________________________________________
//
export const CircleButton = React.forwardRef<HTMLElement, CircleButtonProps>(
  (
    {
      variant = 'ghost',
      size = 'md',
      color = 'primary',
      disabled,
      transition = true,
      tooltip = false,
      children,
      className,
      ...rest
    },
    ref,
  ) => {
    const beforeClasses = clsx(
      'before:opacity-0',
      'before:hover:opacity-100',
      'before:border-b-[7px]',
      'before:border-transparent',
      'before:bottom-auto',
      'before:top-full',
      'before:border-[6px]',
      'before:absolute',
      'before:content-[""]',
      'before:border-b-black',
    );
    const afterClasses = clsx(
      'after:top-full after:mt-3',
      'after:text-xs',
      'after:rounded-md',
      'relative',
      'after:whitespace-nowrap',
      'after:px-2.5',
      'after:py-1',
      'after:bg-black',
      'after:text-white',
      'after:opacity-0',
      'after:hover:opacity-100',
      'after:absolute',
      'after:content-[attr(aria-label)]',
    );

    const classes = clsx(
      'flex',
      'items-center',
      'justify-center',
      'rounded-full',
      'focus:outline-none',
      'focus-visible:ring-2',
      'focus-visible:ring-blue-400',
      variant === 'ghost' && color === 'primary' && 'hover:bg-primary-50',
      variant === 'ghost' && color === 'secondary' && 'hover:bg-secondary-50',
      variant === 'ghost' && color === 'error' && 'hover:bg-error-50',
      disabled && ['pointer-events-none', 'opacity-40'],
      transition && ['transition-all', 'duration-200', 'ease-in'],
      mapSizeClasses[size],
      mapColorClasses[color],
      tooltip && beforeClasses,
      tooltip && afterClasses,
      className,
    );

    // ___________________________________________________________________________
    //
    if (rest.href) {
      return (
        <a
          ref={ref as React.RefObject<HTMLAnchorElement>}
          className={classes}
          {...(disabled && { tabIndex: -1, 'aria-disabled': true })}
          {...(rest as AnchorCircleButtonProps)}
        >
          <div className={clsx(mapIconSize[size])}>{children}</div>
        </a>
      );
    }
    // ___________________________________________________________________________
    //
    return (
      <button
        ref={ref as React.RefObject<HTMLButtonElement>}
        className={classes}
        {...(disabled && { disabled: true, tabIndex: -1 })}
        {...(rest as NativeCircleButtonProps)}
      >
        <div className={clsx(mapIconSize[size])}>{children}</div>
      </button>
    );
  },
);
// ___________________________________________________________________________
//
CircleButton.displayName = 'CircleButton';
