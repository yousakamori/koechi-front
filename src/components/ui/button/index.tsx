export * from './button';
export * from './circle-button';
