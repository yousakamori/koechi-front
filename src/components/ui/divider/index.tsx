import clsx from 'clsx';
import React from 'react';

export type DividerProps = {
  className?: string;
};

export const Divider: React.VFC<DividerProps> = ({ className }) => {
  return <div className={clsx('block', 'border-t', 'border-secondary-200', 'w-full', className)} />;
};
