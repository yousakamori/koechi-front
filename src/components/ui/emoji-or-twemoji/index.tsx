export * from './emoji-or-twemoji';
