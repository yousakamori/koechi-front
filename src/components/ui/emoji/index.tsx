export * from './emoji-picker';
export * from './emoji';
