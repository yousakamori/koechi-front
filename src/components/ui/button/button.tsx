import clsx from 'clsx';
import React from 'react';
import { Spinner } from '@/components/ui/spinner';
// ___________________________________________________________________________
//
const mapVariantClasses = {
  contained: ['shadow', 'text-white', 'hover:shadow-md'],
  outlined: ['shadow', 'border', 'hover:shadow-md'],
  ghost: ['border-none', 'bg-transparent'],
  link: ['hover:underline'],
};

const mapSizeClasses = {
  sm: ['px-3', 'py-1', 'text-sm'],
  md: ['px-4', 'py-2', 'text-sm'],
  lg: ['px-5', 'py-3', 'text-base'],
};

const mapContainedColorClasses = {
  primary: ['bg-primary-600', 'hover:bg-primary-700'],
  secondary: ['bg-secondary-400', 'hover:bg-secondary-500'],
  error: ['bg-error-400', 'hover:bg-error-500'],
};

const mapOutlinedColorClasses = {
  primary: ['bg-white', 'text-primary-500', 'border-primary-500', 'hover:bg-primary-50'],
  secondary: ['bg-white', 'text-secondary-500', 'border-secondary-300', 'hover:bg-secondary-50'],
  error: ['bg-white', 'text-error-500', 'border-error-500', 'hover:bg-error-50'],
};

const mapGhostColorClasses = {
  primary: ['text-primary-500', 'hover:bg-primary-100'],
  secondary: ['text-secondary-500', 'hover:bg-secondary-100'],
  error: ['text-error-500', 'hover:bg-error-100'],
};

const mapLinkColorClasses = {
  primary: ['text-primary-500'],
  secondary: ['text-secondary-500'],
  error: ['text-error-500'],
};
// ___________________________________________________________________________
//
export type BaseButtonProps = {
  variant?: keyof typeof mapVariantClasses;
  size?: keyof typeof mapSizeClasses;
  color?: keyof typeof mapContainedColorClasses;
  loading?: boolean;
  disabled?: boolean;
  fullWidth?: boolean;
  roundedFull?: boolean;
  transition?: boolean;
};

export type AnchorButtonProps = BaseButtonProps &
  Omit<React.AnchorHTMLAttributes<HTMLAnchorElement>, keyof BaseButtonProps>;

export type NativeButtonProps = BaseButtonProps &
  Omit<React.ButtonHTMLAttributes<HTMLButtonElement>, keyof BaseButtonProps>;

export type ButtonProps = AnchorButtonProps & NativeButtonProps;
// ___________________________________________________________________________
//
export const Button = React.forwardRef<HTMLElement, ButtonProps>(
  (
    {
      variant = 'contained',
      color = 'primary',
      size = 'md',
      loading = false,
      disabled = false,
      fullWidth = false,
      roundedFull = false,
      transition = true,
      children,
      className,
      ...rest
    },
    ref,
  ) => {
    const classes = clsx(
      'inline-flex',
      'items-center',
      'justify-center',
      'focus:outline-none',
      'focus-visible:ring-2',
      'focus-visible:ring-blue-400',
      fullWidth && 'w-full',
      roundedFull ? 'rounded-full' : 'rounded-lg',
      (disabled || loading) && ['pointer-events-none', 'opacity-40'],
      transition && ['transition-all', 'duration-200', 'ease-in'],
      variant === 'contained' && mapContainedColorClasses[color],
      variant === 'outlined' && mapOutlinedColorClasses[color],
      variant === 'ghost' && mapGhostColorClasses[color],
      variant === 'link' && mapLinkColorClasses[color],
      mapVariantClasses[variant],
      mapSizeClasses[size],
      className,
    );
    // ___________________________________________________________________________
    //
    if (rest.href) {
      return (
        <a
          ref={ref as React.RefObject<HTMLAnchorElement>}
          className={classes}
          {...((disabled || loading) && {
            tabIndex: -1,
            'aria-disabled': true,
          })}
          {...(rest as AnchorButtonProps)}
        >
          {loading && <Spinner className='mr-2' color='secondary' size='sm' />}
          {children}
        </a>
      );
    }
    // ___________________________________________________________________________
    //
    return (
      <button
        ref={ref as React.RefObject<HTMLButtonElement>}
        className={classes}
        {...((disabled || loading) && { disabled: true, tabIndex: -1 })}
        {...(rest as NativeButtonProps)}
      >
        {loading && <Spinner className='mr-2' color='secondary' size='sm' />}
        {children}
      </button>
    );
  },
);
// ___________________________________________________________________________
//
Button.displayName = 'Button';
