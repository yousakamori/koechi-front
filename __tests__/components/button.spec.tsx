import '@testing-library/jest-dom';

// import { render, screen } from "@testing-library/react";
import { render } from '@testing-library/react';

import { Button } from '@/components/ui/button';

// import { Button } from "@/components/ui";

// import { Button } from "@/components/ui";
// import { Button } from "../index";
// import { Button } from "../../components/ui";

describe('Button', () => {
  it('Should render hello text', () => {
    const { getByText } = render(<Button>Click me</Button>);
    const text = getByText('Click me');
    expect(text).toBeVisible();
  });

  it('should render a disabled button if loading prop is passed', () => {
    const { getByText } = render(<Button loading={true}>Click me</Button>);
    const text = getByText('Click me');

    expect(text.closest('button')).toBeDisabled();
  });

  it('should render a disabled button if disabled prop is passed', () => {
    const { getByText } = render(<Button disabled={true}>Click me</Button>);
    const text = getByText('Click me');

    expect(text.closest('button')).toBeDisabled();
  });

  test('primary', () => {
    const { container } = render(
      <Button color='primary' size='md'>
        Primary
      </Button>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('primary disabled', () => {
    const { container } = render(
      <Button color='primary' size='md' disabled>
        Primary
      </Button>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('primary large', () => {
    const { container } = render(
      <Button color='primary' size='lg'>
        Primary
      </Button>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('secondary', () => {
    const { container } = render(
      <Button color='secondary' size='md'>
        Secondary
      </Button>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('secondary disabled', () => {
    const { container } = render(
      <Button color='secondary' size='md' disabled>
        Secondary
      </Button>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
