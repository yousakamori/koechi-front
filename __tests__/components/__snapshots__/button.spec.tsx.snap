// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button primary 1`] = `
<button
  class="flex items-center justify-center outline-none rounded-lg focus:outline-none focus-visible:ring-2 focus-visible:ring-blue-400 transition-all duration-200 ease-in shadow text-white hover:shadow-md px-4 py-2 text-sm bg-primary-400 hover:bg-primary-500 active:bg-primary-600"
>
  Primary
</button>
`;

exports[`Button primary disabled 1`] = `
<button
  class="flex items-center justify-center outline-none rounded-lg focus:outline-none focus-visible:ring-2 focus-visible:ring-blue-400 pointer-events-none opacity-40 transition-all duration-200 ease-in shadow text-white hover:shadow-md px-4 py-2 text-sm bg-primary-400 hover:bg-primary-500 active:bg-primary-600"
  disabled=""
  tabindex="-1"
>
  Primary
</button>
`;

exports[`Button primary large 1`] = `
<button
  class="flex items-center justify-center outline-none rounded-lg focus:outline-none focus-visible:ring-2 focus-visible:ring-blue-400 transition-all duration-200 ease-in shadow text-white hover:shadow-md px-5 py-3 text-md bg-primary-400 hover:bg-primary-500 active:bg-primary-600"
>
  Primary
</button>
`;

exports[`Button secondary 1`] = `
<button
  class="flex items-center justify-center outline-none rounded-lg focus:outline-none focus-visible:ring-2 focus-visible:ring-blue-400 transition-all duration-200 ease-in shadow text-white hover:shadow-md px-4 py-2 text-sm bg-secondary-400 hover:bg-secondary-500 active:bg-secondary-600"
>
  Secondary
</button>
`;

exports[`Button secondary disabled 1`] = `
<button
  class="flex items-center justify-center outline-none rounded-lg focus:outline-none focus-visible:ring-2 focus-visible:ring-blue-400 pointer-events-none opacity-40 transition-all duration-200 ease-in shadow text-white hover:shadow-md px-4 py-2 text-sm bg-secondary-400 hover:bg-secondary-500 active:bg-secondary-600"
  disabled=""
  tabindex="-1"
>
  Secondary
</button>
`;
